ADALiOS/ADALiOS/ADTokenCache.m

```diff
@@ -0,0 +1,546 @@
+            CHECK_ERROR([userId isKindOfClass:[NSString class]] || [userId isKindOfClass:[NSNull class]], AD_ERROR_BAD_CACHE_FORMAT, @"User ID key not the expected class type");
+            id userDict = [tokens objectForKey:userId];
+            CHECK_ERROR([userDict isKindOfClass:[NSMutableDictionary class]], AD_ERROR_BAD_CACHE_FORMAT, @"User ID should have mutable dictionaries in the cache");
+            
+            for (id adkey in userDict)
+            {
+                // On the first level we're expecting NSDictionaries keyed off of ADTokenCacheStoreKey
+                CHECK_ERROR([adkey isKindOfClass:[ADTokenCacheKey class]], AD_ERROR_BAD_CACHE_FORMAT, @"Key is not the expected class");
+                id token = [userDict objectForKey:adkey];
+                CHECK_ERROR([token isKindOfClass:[ADTokenCacheItem class]], AD_ERROR_BAD_CACHE_FORMAT, @"Token is not of expected class type!");
+            }
+        }
+    }
+    
+    
+    NSDictionary* idtokens = [cache objectForKey:@"idtokens"];
+    if (idtokens)
+    {
+        CHECK_ERROR([idtokens isKindOfClass:[NSMutableDictionary class]], AD_ERROR_BAD_CACHE_FORMAT, @"idtokens is not a dictionary!");
+        for (id idtoken_key in idtokens)
+        {
+            CHECK_ERROR([idtoken_key isKindOfClass:[NSString class]], AD_ERROR_BAD_CACHE_FORMAT, @"idtoken key is not a string!");
+            id idtoken = [idtokens objectForKey:idtoken_key];
+            CHECK_ERROR([idtoken isKindOfClass:[NSString class]], AD_ERROR_BAD_CACHE_FORMAT, @"idtoken is not a string!");
+        }
+    }
+    
+    return YES;
+}
+
+#pragma mark -
+#pragma mark ADTokenCacheAccessor Protocol Implementation
+
+/*! May return nil, if no cache item corresponds to the requested key
+ @param key: The key of the item.
+ @param user: The specific user whose item is needed. May be nil, in which
+ case the item for the first user in the cache will be returned.
+ @param error: Will be set only in case of ambiguity. E.g. if userId is nil
+ and we have tokens from multiple users. If the cache item is not present,
+ the error will not be set. */
+- (ADTokenCacheItem *)getItemWithKey:(ADTokenCacheKey *)key
+                                   userId:(NSString *)userId
+                                    error:(ADAuthenticationError * __autoreleasing *)error
+{
+    NSArray<ADTokenCacheItem *> * items = [self getItemsWithKey:key userId:userId error:error];
+    if (!items || items.count == 0)
+    {
+        return nil;
+    }
+    
+    if (items.count == 1)
+    {
+        return items.firstObject;
+    }
+    
+    ADAuthenticationError* adError =
+    [ADAuthenticationError errorFromAuthenticationError:AD_ERROR_MULTIPLE_USERS
+                                           protocolCode:nil
+                                           errorDetails:@"The token cache store for this resource contain more than one user. Please set the 'userId' parameter to determine which one to be used."];
+    if (error)
```

Comment:
This if (error) isn't checking the getItemsWithKey call. That call is being checked with if (!items) right after the call.

This if (error) is checking to see if the developer supplied the optional error parameter to make sure I don't segfault if it wasn't provided.